### PR TITLE
Abstract sched

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
## Purpose / Description
The initial problem that I had was some automated test having unexpected result. Mostly because there is an empty constructor in Sched. And that actually, what it really meant is that SchedV2 should never have inherited from Sched. Indeed, it reuses absolutly no code and no variable from Sched, and it's not semantically a specific kind of Sched.

It also corrected the name of SchedV2, so that it's the same as in Anki. Which changes nothing as this name is never used here nor there.

## Approach
I simply created an abstract Sched class, from which both Sched and SchedV2 inherits. This abstract class declare as public abstract all public classes from Sched that are also in SchedV2. This way, we have in a proper java way what we already had in Python: no inheritance between both classes, but being able to use the same class name for both variables

I also did change to private some methods which were in one class and not the other, and which were never used outside of the Sched class. 

## How Has This Been Tested?
Started Ankidroid and checked that reviewing still happened correctly. 

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by Arthur-Milchior** [comment]
Codacy reviews shows problems which were already in the original code. Since those are not new errors, I'll leave them

**Comment by mikehardy** [comment]
For Codacy:
The unused imports are new: https://app.codacy.com/manual/AnkiDroid/Anki-Android_2/pullRequest?prid=4449466#possible-new-issues

I'll see if I can add an optional '_' to the method name acceptable filter, it's clearly part of our standards

**Comment by Arthur-Milchior** [comment]
I thought i did correct the unused import shown the first time. Sorry. I'll
edit it quickly
Can i run codacy directly on my computer without doing a PR?

**Comment by Arthur-Milchior** [comment]
Pushed again without unused import in Abstract Sched

**Comment by mikehardy** [comment]
> I thought i did correct the unused import shown the first time. Sorry. I'll edit it quickly Can i run codacy directly on my computer without doing a PR? 

You can, I built a script that does it.

It requires some setup: https://github.com/ankidroid/Anki-Android/blob/master/tools/quality-check/codacy-ankidroid.sh

https://github.com/ankidroid/Anki-Android/blob/690dd7b60f99e25b9dbf69a4d5a18759af9b17f2/tools/quality-check/codacy-ankidroid.sh#L3-L6

Docker as a technology is a big install, and the codacy analysis Docker container is some tens of MB. Docker (and containerization in general) is at least pretty amazing technology, it took over the world for a reason and I'm happy to have some exposure to it.

Then it does the analysis inside the docker container - only takes a few seconds on a big machine. On a smaller machine it could be a little heavy but still way better than waiting for codacy.com to send you email.

When I'm developing I use it all the time. I do like this:

```
git co master
./bin/codacy-ankidroid.sh master  # the script copy in bin has the API token in it
git co issue5011-4 # or whatever your work branch is
./bin/codacy-ankidroid.sh issue5011-4
diff ../reports/codacy/codacy-report-master.txt ../reports/codacy/codacy-report-issue5011-4.txt 
```

Then I sift through the diff blocks and re-write git history until I've fixed things - normally the diff blocks are matching sets of lines except for line number changes, and the actual diffs stand out pretty well.

I can get you the API token if you want to try it, and if you find it useful I'll add something to the wiki

File: AnkiDroid/src/main/java/com/ichi2/libanki/SchedV2.java [hunk 1]
```diff
@@ -58,14 +58,14 @@
...
-public class SchedV2 extends Sched {
+public class SchedV2 extends AbstractSched {
 
 
 
     // Not in libanki
     private static final int[] FACTOR_ADDITION_VALUES = { -150, 0, 150 };
 
-    private String mName = "std";
+    private String mName = "std2";
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/SchedV2.java]: I was curious if this was a semantic change as it is used in the public method getName(), but in AndroidStudio there are apparently zero usages of getName() :man_shrugging:

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/SchedV2.java]: and yes you mentioned that in your comment but I double-checked :-)

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/SchedV2.java]: Name's value is never used in Anki either

**Comment by Arthur-Milchior** [comment]
I would like to note that there was an unexpected change while this PR was reviewed. E.g. between November and April.
In anki, the code of scheduler is factorized. Scheduler inherits from scheduler V2, and only redefines what differs between them.

I believe this PR is still useful. Because having scheduler inheriting scheduler v2 is horrible as far as typing goes. It makes no sens. Having an abstract parent class and both inheriting from it makes far more sens. And typing is more important in java than in python, so I'm still happy I did this PR.

But this lead to a second question. Do we decide to imitates Anki and asks for sched to inherits SchedV2 ? 
Do we let things stay as they are right now ?
Or would you accept what I think is the best solution: factorize the code by putting duplicate code into the abstract scheduler, and keeping in sched and schedv2 only what differs. This will make the code slightly different than anki's code, but far more easy to read

**Comment by mikehardy** [comment]
I think having it stay with upstream is valuable - I refer to the upstream code all the time and the fidelity between them is a big help for me

**Pull request merged**